### PR TITLE
Make `app_commands.Command` and others comparable with `AppCommand(Group)`

### LINKED ISSUE #7669: compare AppCommand and Command
### Summary

Add posibility to compare `AppCommand` - API object and `Command` - runtime one

### What is the feature request for?

The core library

### The Problem

It would be easier to determine if command synchronization is required.

### The Ideal Solution

```py
synced_commands = await bot.tree.fetch_commands()
existing_commands = bot.tree.get_commands()

if synced_commands != existing_commands:
    await bot.tree.sync()
```

### The Current Solution

In turn, compare first the type of command, then the options if any.

### Additional Context

_No response_

### PR DESCRIPTION
**Pull request by Soheab**
## Summary

Fixes #7669 and resolves [card-81287257](https://github.com/Rapptz/discord.py/projects/3#card-81287257)

This PR makes it possible to compare the following classes to the `AppCommand(Group)` modals and itself:

- app_commands.Command
- app_commands.ContextMenu
- app_commands.Group

I renamed the `AppCommand`'s `.to_dict()` method to `.to_comparison_dict()` [as discussed](https://discord.com/channels/336642139381301249/603069307286454290/984789577602592829) in the [discord.py server](https://discord.com/invite/dpy)

I have tested this thoughtfully but not with every possibility but it should be fine...

## Checklist

- [x] If code changes were made then they have been tested.
    - [] I have updated the documentation to reflect the changes.
- [x] This PR fixes an issue.
- [] This PR adds something new (e.g. new method or parameters).
- [x] This PR is a breaking change (e.g. methods or parameters removed/renamed)
- [ ] This PR is **not** a code change (e.g. documentation, README, ...)

### CODE REVIEW

File: discord/app_commands/commands.py [hunk 1]
```diff
@@ -534,6 +534,26 @@ def __init__(
...
+            if self._guild_ids is not None:
+                cls_to_dict['dm_permission'] = True
+
+            return cls_to_dict == other.to_comparison_dict()
+        else:
+            return False
+
+    def __ne__(self, other: object) -> bool:
+        return not self.__eq__(other)
+
```
Comment (Rapptz) [inline discord/app_commands/commands.py:556]: This is already implied

```suggestion
```

File: discord/app_commands/commands.py [hunk 2]
```diff
@@ -972,6 +992,17 @@ def __init__(
...
+        if isinstance(other, self.__class__):
+            return self.to_dict() == other.to_dict()
+        elif isinstance(other, AppCommand):
+            return self.to_dict() == other.to_comparison_dict()
+        else:
+            return False
+
+    def __ne__(self, other: object) -> bool:
+        return not self.__eq__(other)
+
```
Comment (Rapptz) [inline discord/app_commands/commands.py:1005]: This is already implied

```suggestion
```

File: discord/app_commands/models.py [hunk 3]
```diff
@@ -201,15 +201,22 @@ def _from_data(self, data: ApplicationCommandPayload) -> None:
...
             'options': [opt.to_dict() for opt in self.options],
-        }  # type: ignore # Type checker does not understand this literal.
+            'nsfw': self.nsfw,
+            'default_member_permissions': self.default_member_permissions and self.default_member_permissions.value,
+            'dm_permission': self.dm_permission,
+        }
+        # context menus don't support this.
+        if self.type in (AppCommandType.user, AppCommandType.message):
+            del base['options']
+            del base['description']
```
Comment (Rapptz) [inline discord/app_commands/models.py:217]: Invert the condition so keys are added rather than removed.

File: discord/app_commands/models.py [hunk 4]
```diff
@@ -201,15 +201,22 @@ def _from_data(self, data: ApplicationCommandPayload) -> None:
         self.dm_permission: bool = dm_permission
         self.nsfw: bool = data.get('nsfw', False)
 
-    def to_dict(self) -> ApplicationCommandPayload:
-        return {
-            'id': self.id,
+    def to_comparison_dict(self) -> ApplicationCommandPayload:
```
Comment (Rapptz) [inline discord/app_commands/models.py:204]:
```suggestion
    def to_comparison_dict(self) -> Dict[str, Any]:
```

File: discord/app_commands/commands.py [hunk 5]
```diff
@@ -1249,6 +1280,17 @@ def __init__(
...
+    def __eq__(self, other: object) -> bool:
+        if isinstance(other, self.__class__):
+            return self.to_dict() == other.to_dict()
+        elif isinstance(other, AppCommand):
+            return self.to_dict() == other.to_comparison_dict()
+        else:
+            return False
+
+    def __ne__(self, other: object) -> bool:
+        return not self.__eq__(other)
```
Comment (Rapptz) [inline discord/app_commands/commands.py:1292]:
```suggestion
```

**Review by Rapptz — CHANGES_REQUESTED** [review]
Note that `api_model == user_command` will fail. You need to make sure the `__eq__` and `__ne__` of the app command models return `NotImplemented` when the types differ so it can propagate it back to the user facing types.

That aside, I'm a bit worried about tying the comparison behaviour to `to_dict()`, especially when localisations come into play.

**Comment by Soheab** [comment]
> Note that `api_model == user_command` will fail. You need to make sure the `__eq__` and `__ne__` of the app command models return `NotImplemented` when the types differ so it can propagate it back to the user facing types.

Done, thanks.

> That aside, I'm a bit worried about tying the comparison behaviour to `to_dict()`, especially when localisations come into play.

Doing it with a dict seems like the easiest way IMO as it includes all the values we need to compare and localisations stuff should be the same afaik.

> Also note that by defining `__eq__` you break the hashability of the objects since now there's no `__hash__`, causing test failures.

Fixed. There are still some tests failing but I don't think those are in the scope of this PR?

**Comment by Rapptz** [comment]
As discussed in the server, I think this feature has kind of become infeasible due to localisation. The original request was also kind of dubious since the server would know nothing changed making the sync call have no side effects.

**Pull request closed without merging**